include/SFML/Window/Event.hpp

```diff
@@ -34,195 +34,350 @@
     };
+
+private:
+    using VariantType = std::variant<
+        Empty,
+        Closed,
+        Resized,
+        FocusLost,
+        FocusGained,
+        TextEntered,
+        KeyPressed,
+        KeyReleased,
+        MouseWheelScrolled,
+        MouseButtonPressed,
+        MouseButtonReleased,
+        MouseMoved,
+        MouseEntered,
+        MouseLeft,
+        JoystickButtonPressed,
+        JoystickButtonReleased,
+        JoystickMoved,
+        JoystickConnected,
+        JoystickDisconnected,
+        TouchBegan,
+        TouchMoved,
+        TouchEnded,
+        SensorChanged>;
+
+    template <typename T, typename... Ts>
+    static constexpr bool isInParameterPack(const std::variant<Ts...>&)
+    {
+        return (std::is_same_v<T, Ts> || ...);
+    }
+
+    template <typename T>
+    static constexpr bool isEventType = isInParameterPack<T>(VariantType());
+
+public:
+    ////////////////////////////////////////////////////////////
+    /// \brief Default constructor
+    ///
+    /// Sets the event to Type::Empty
+    ///
+    ////////////////////////////////////////////////////////////
+    Event() = default;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Construct from a given sf::Event subtype
+    ///
+    /// \param t Event subtype
+    ///
+    ////////////////////////////////////////////////////////////
+    template <typename T>
+    Event(const T& t);
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Get current event type
+    ///
+    /// \return Current event type
+    ///
+    ////////////////////////////////////////////////////////////
+    [[nodiscard]] Type getType() const;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Check current event type
+    ///
+    /// \return True if template parameter is current event type
+    ///
+    ////////////////////////////////////////////////////////////
+    template <typename T>
+    [[nodiscard]] bool is() const;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Get particular event type
+    ///
+    /// Throws std::bad_variant_access if event type does not
+    /// match the current event type.
+    ///
+    /// \return Reference to current event type
+    ///
+    ////////////////////////////////////////////////////////////
+    template <typename T>
+    [[nodiscard]] const T& get() const;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Get particular event type
+    ///
+    /// \return Address of current event type, otherwise nullptr
+    ///
+    ////////////////////////////////////////////////////////////
+    template <typename T>
+    [[nodiscard]] const T* getIf() const;
+
+private:
+    ////////////////////////////////////////////////////////////
+    // Member data
+    ////////////////////////////////////////////////////////////
+    VariantType m_data; //!< Event data
```

Comment:
At the risk of losing some encapsulation, it would be nice to have a public getter that returns the raw `VariantType&` to the user. This would allow users to use `std::visit` if they so desire. Could be a separate PR, though.

Reply:
See the PR description for why I don't want to add that.

Reply:
This?

> You must visit a variant with the perfect set of callbacks for the elements in the variant. If we allow users to visit this variant then we're effectively preventing ourselves from adding new events in a given major release. For now I'd prefer to retain that ability. We can always add this later if desired.

I'm not sure I get it. If a user decides to use `std::visit` with an exhaustive set of callbacks, they are intentionally saying *"please break my build if a new event is added, because I care about exhaustiveness"*. They are using `std::visit` exactly to get notified that something new was added.

Note that a user can still use a catch-all callback in a `std::visit` if they don't care about future-proof exhaustiveness:

```cpp
std::visit(overload{
    [](const A&) { ... },
    [](const B&) { ... },
    [](const auto&) { ...fallback... },
})
```

Reply:
My impression was that visiting a variant requires overloads for all types. Some quick tests confirms that suspicion of mine. What I think I was missing was your `const auto&` overload. It's basically a `default:` case in a switch statement. Nobody would say that adding a new value to an enum constitutes an API break just because someone chose to write a `switch` without a `default:` case.

This may suffice to assuage my concerns about supporting `std::visit` since we can point to this if and when a user complains that a new event type causes their program to stop compiling.

Reply:
https://github.com/SFML/SFML/compare/event_api...visit_event

Here's what that would look like

Reply:
It's not just a forward-compatibility hazard in terms of adding new events, but at the off chance we'd ever use a different implementation than `std::variant`. We would lock ourselves into this specific implementation, which begs the question of why having an abstraction in the first place.

I would expose the internal variant only if there is concrete demand, not prematurely. It's easy to add this, hard to remove. And if there is a concrete, real-world use case, we can also try to understand the usage pattern better and consider if the existing API can be improved somehow.

Reply:
Agreed. I'm glad to know that supporting `std::visit` is in fact feasible but I'd like to see more proof of usefulness before we expose such an implementation detail. As demonstrated this is really easy to add at any point in the future.